services/src/contexts/migrations/database_migration.rs

```diff
@@ -0,0 +1,300 @@
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    let stmt = match conn
+        .prepare("SELECT database_version from geoengine;")
+        .await
+    {
+        Ok(stmt) => stmt,
+        Err(e) => {
+            if let Some(code) = e.code() {
+                if *code == SqlState::UNDEFINED_TABLE {
+                    return Ok(None);
+                }
+            }
+            return Err(crate::error::Error::TokioPostgres { source: e });
+        }
+    };
+
+    let row = conn.query_one(&stmt, &[]).await?;
+
+    Ok(Some(row.get(0)))
+}
+
+#[derive(Debug, PartialEq)]
+pub enum MigrationResult {
+    CreatedDatabase,
+    MigratedDatabase,
+    AlreadyUpToDate,
+}
+
+/// Migrate the database to the latest version. If the database is empty, the initial migration is applied.
+///
+/// # Panics
+///
+/// Panics if there is an error in the migration logic.
+pub async fn migrate_database<Tls>(
+    conn: &mut PooledConnection<'_, PostgresConnectionManager<Tls>>,
+    migrations: &[Box<dyn Migration>],
+) -> Result<MigrationResult>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    let pre_migration_version = determine_current_database_version(conn).await?;
+    info!("Current database version: {:?}", pre_migration_version);
+
+    // start with the first migration after the current version
+    let applicable_migrations = migrations
+        .iter()
+        .skip_while(|m| m.prev_version() != pre_migration_version);
+
+    let mut current_version = pre_migration_version.clone();
+
+    for migration in applicable_migrations {
+        ensure!(
+            migration.prev_version() == current_version,
+            UnexpectedDatabaseVersionDuringMigration {
+                expected: migration.prev_version().unwrap_or_default(),
+                found: current_version.unwrap_or_default()
+            }
+        );
+
+        info!("Applying migration: {}", migration.version());
+
+        let tx = conn.build_transaction().start().await?;
+
+        migration.migrate(&tx).await?;
+
+        let stmt = tx
+            .prepare("UPDATE geoengine SET database_version = $1")
+            .await?;
+
+        tx.execute(&stmt, &[&migration.version()]).await?;
+
+        tx.commit().await?;
+
+        current_version = Some(migration.version());
+    }
+
+    let current_version = determine_current_database_version(conn)
+        .await?
+        .expect("after migration, there should be a current version.");
+
+    let latest_version = migrations
+        .last()
+        .expect("there should be at least one migration, namely the inital migration.")
+        .version();
```

Comment:
Can't we do this here instead of just checking it here?

Reply:
do what exactly?

Reply:
update the version instead of determining that the version is correct